### PR TITLE
Add Java helper code for establishing binder transport connection

### PR DESCRIPTION
**Pull request by sifmelcara**
These Java code will be invoked by binder transport C++ implementation
through JNI to establish the connection between client and server.

The code is locally tested with other pending changes to make sure it
works correctly.

For now we only make sure it builds in CI. We will port proper tests
from internal repository later.

### CODE REVIEW

**Comment by sifmelcara** [comment]
cc @TaWeiTu @napathome 

Will ask for review from Nicolas after you guys' review

File: src/core/ext/transport/binder/java/io/grpc/binder/cpp/NativeConnectionHelper.java
Comment (napathome) [inline src/core/ext/transport/binder/java/io/grpc/binder/cpp/NativeConnectionHelper.java:21]: Could you document how user should use this helper class? For example, after user call tryEstablishConnection, when should user call getServiceBinder?

Comment (napathome) [reply src/core/ext/transport/binder/java/io/grpc/binder/cpp/NativeConnectionHelper.java:21]: Since tryEstablishConnection could fail, should we return something to let user know?

File: src/core/ext/transport/binder/java/io/grpc/binder/cpp/SyncServiceConnection.java
Comment (napathome) [inline src/core/ext/transport/binder/java/io/grpc/binder/cpp/SyncServiceConnection.java:68]: Should we return Optional here for the case service is disconnected?

Comment (sifmelcara) [reply src/core/ext/transport/binder/java/io/grpc/binder/cpp/SyncServiceConnection.java:68]: Added a TODO in NativeConnectionHelper

File: src/core/ext/transport/binder/java/io/grpc/binder/cpp/NativeConnectionHelper.java
Comment (sifmelcara) [reply src/core/ext/transport/binder/java/io/grpc/binder/cpp/NativeConnectionHelper.java]: > Could you document how user should use this helper class? For example, after user call tryEstablishConnection, when should user call getServiceBinder?

Added some comments. Actually user will not call this class. This class is only invoked from our implementation in gRPC

> Since tryEstablishConnection could fail, should we return something to let user know?

Added a TODO. For now we just assume it always success in our prototype.

**Review by napathome — APPROVED** [review]

**Comment by sifmelcara** [comment]
Note: Before merging this, we will need to add a simple replace rule to copy.bara.sky to replace `@build_bazel_rules_android//` in the BUILD file in this CL

**Comment by ctiller** [comment]
Can you talk about what the new WORKSPACE file is for?

**Comment by sifmelcara** [comment]
Since this specific BUILD file will need Android SDK/NDK properly configured to build, we put it into a `local_repository` so that the `...` expansion in bazel will exclude this BUILD file.

Without this `local_repository` trick, scripts like `gen_compilation_database.py` will fail because it will consider the `load` statement in the BUILD file a failure.

Other than the purpose of avoiding bazel failure, I think creating local_repository makes some sense because it is Android only Java code (and will be compiled separately) which we want to avoid it interfere with the rest of gRPC core.

(Oops I thought I documented in the WORKSPACE file but the change is lost, will add it back if this approach is ok)

**Comment by ctiller** [comment]
I'm going to leave the ok question up to @nicolasnoble but yeah let's at least add some documentation here.

**Comment by nicolasnoble** [comment]
Hah, yeah, that's a good trick to avoid the `...` breakage, but I'm not 100% sure about how this will look on import. We should trigger the test for it.

**Comment by sifmelcara** [comment]
We have done a copybara import and it looks good. I believe the WORKSPACE files in grpc directory are not used in google's internal repository. On top of that, in our internal repository people don't need to configure Android SDK because it is already configured in the source tree.

I guess let's merge and see? We can always revert this if we find out non trivial issue during daily copybara import

**Review by nicolasnoble — APPROVED** [review]

**Pull request merged**
